file-format-spec.ts

```diff
@@ -0,0 +1,58 @@
+// This file contains types which specify the speedscope file format.
+
+export interface SerializedFrame {
+  name: string
+  file?: string
+  line?: number
+  col?: number
+}
+
+export interface SerializedNode {
+  // Index into the frames array on the SerializedProfile
+  frame: number
+
+  // Index into the nodes array on the SerializedProfile
+  parent?: number
+}
+
+export type WeightUnit =
+  | 'none'
+  | 'nanoseconds'
+  | 'microseconds'
+  | 'milliseconds'
+  | 'seconds'
+  | 'bytes'
+
+export interface SerializedSamplingProfile {
+  // Type of profile. This will future proof the file format to allow many
+  // different kinds of profiles to be contained and each type to be part of
+  // a discriminate union.
+  type: 'SamplingProfile'
+
+  // Name of the profile. Typically a filename for the source of the profile.
+  name: string
+
+  // List of all call frames
+  frames: SerializedFrame[]
+
+  // List of nodes in the call tree
+  nodes: SerializedNode[]
+
+  // List of indices into nodes, with -1 indicating that the call-stack
+  // was empty at the time of the sample
+  samples: number[]
```

Comment:
Would it be possible to have an empty samples array for trace sources that are not using sampling?

Reply:
Speedscope currently imports from a wide variety of sampling profile formats (Chrome, Firefox, Instruments, Ruby stackprof, and anything that exports in the collapsed stack format). So this isn't just to export instruments files, but to export any sampling profile.

That said, your comment does suggest to me that there may be a better file format that would seamlessly support both sampled profiles & instrumented profiles. Let me mull it over and get back to you.

Thanks for the questions!